# Введение в анализ данных
## НИУ ВШЭ, 2019-2020 учебный год

### Домашнее задание №3

Задание выполнил(а): Разуваев Никита Сергеевич

### Общая информация
Дата выдачи: 09.04.2020

Дедлайн: 24.04.2020 23:59 MSK

### О задании

В этом домашнем задании вы будете работать с линейной классификацией, попрактикуетесь на реальной задаче классификации текстов.

Для решения этого домашнего задания намного удобнее будет использовать Colab, так как данных много.

### Оценивание и штрафы

За сдачу задания позже срока на итоговую оценку за задание накладывается штраф в размере 1 балл в день, но получить отрицательную оценку нельзя.

__Внимание!__ Домашнее задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов.

### Формат сдачи
Загрузка файлов с решениями происходит в системе Anytask.

### Подготовка данных

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

from tqdm.notebook import tqdm

In [2]:
# Датасет можно скачать здесь

!wget https://www.dropbox.com/s/tg55q9mrziroyrs/train_subset.csv

"wget" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.


### Данные

Мы имеем дело с данными с торговой платформы Avito.
Для каждого товара представлены следующие параметры:
 - title
 - description
 - Category_name
 - Category

Имеется информация об объектах 50 классов.
Задача: по новым объектам (title, description) предсказать Category.
(Очевидно, что параметр Category_name для предсказания классов использовать нельзя)

In [3]:
data = pd.read_csv("train_subset.csv", index_col='id')

data.head()

title  \
id                                   
382220                    Прихожая   
397529   Кордиант 215/55/16 Летние   
584569                        Стол   
2513100                 Комбинезон   
1091886                   Ветровка   

                                               description  \
id                                                           
382220                           В хорошем состоянии. Торг   
397529   Кордиант 215/55/16 Летние/\n /\nАртикул: 1737l...   
584569   Стол, 2 рабочих места . Стол серого цвета, в д...   
2513100                                       Размер-42/44   
1091886                                          На 2 года   

                     Category_name  Category  
id                                            
382220           Мебель и интерьер        20  
397529       Запчасти и аксессуары        10  
584569           Мебель и интерьер        20  
2513100  Одежда, обувь, аксессуары        27  
1091886     Детская одежда и обувь        29

In [4]:
data.shape

(30000, 4)

In [5]:
X = data[['title', 'description']].to_numpy()
y = data['Category'].to_numpy()

del data

Сразу разделим выборку на train и test.
Никакие данные из test для обучения использовать нельзя!

In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [7]:
X_train[:5]

array([['Сапоги 46 размер новые', 'Сапоги 46 размер новые'],
       ['Светильники потолочный swarovski',
        'светильники потолочные swarovski 6 штук , цена за штуку. В эксплуатации 2 года , продаются в связи со сменой интерьера в квартире'],
       ['iPhone 7 plus 128GB Red красный в наличии',
        '\xa0/\n/\n Данная цена только для подписчиков Instagram: iQmac/\n/\n Новый красный айфон 7 Plus в наличии это элегантный и мощный смартфон, который готов в полной мере раскрыть новые возможности iOS 10. Аппарат с 4-ядерным процессором А10 и 3 ГБ ОЗУ с легкостью решает самые ресурсоемкие задачи, позволяя наслаждаться быстродействием «тяжелых» приложений и игр на 5,5-дюймовом дисплее. Аппарат получил экран, как у iPad Pro, так что картинка теперь соответствует кинематографическому стандарту.'],
       ['Пион Ирис Ромашка рассада',
        'Пион куст 500 р ( более 10 шт)/\nСаженец/ корень 100р/\nРастут у нас более 70 лет/\nРозовые, бордовые и белые/\nНа фото цветы 2018г/\nП. Зубчанинов

### Токенизация (1 балл)


Токенизация -- разбиение текста на мелкие части, которые можно обработать машинными методами.
Можно использовать разные алгоритмы токенизации.
Давайте пока остановимся на простом WordPunctTokenizer.


In [8]:
from nltk.tokenize import WordPunctTokenizer


tokenizer = WordPunctTokenizer()


def preprocess(text: str) -> str:
    return ' '.join(tokenizer.tokenize(text.lower()))


text = 'Здраствуйте. Я, Кирилл. Хотел бы чтобы вы сделали игру, 3Д-экшон суть такова...'
print("before:", text,)
print("after:", preprocess(text),)

before: Здраствуйте. Я, Кирилл. Хотел бы чтобы вы сделали игру, 3Д-экшон суть такова...
after: здраствуйте . я , кирилл . хотел бы чтобы вы сделали игру , 3д - экшон суть такова ...


__Задание:__ Токенизируйте title и description в train и test

In [9]:
# Your code here
preprocess_func = np.vectorize(preprocess)
X_train = preprocess_func(X_train)

In [10]:
X_train[:10]

array([['сапоги 46 размер новые', 'сапоги 46 размер новые'],
       ['светильники потолочный swarovski',
        'светильники потолочные swarovski 6 штук , цена за штуку . в эксплуатации 2 года , продаются в связи со сменой интерьера в квартире'],
       ['iphone 7 plus 128gb red красный в наличии',
        '/ / данная цена только для подписчиков instagram : iqmac / / новый красный айфон 7 plus в наличии это элегантный и мощный смартфон , который готов в полной мере раскрыть новые возможности ios 10 . аппарат с 4 - ядерным процессором а10 и 3 гб озу с легкостью решает самые ресурсоемкие задачи , позволяя наслаждаться быстродействием « тяжелых » приложений и игр на 5 , 5 - дюймовом дисплее . аппарат получил экран , как у ipad pro , так что картинка теперь соответствует кинематографическому стандарту .'],
       ['пион ирис ромашка рассада',
        'пион куст 500 р ( более 10 шт )/ саженец / корень 100р / растут у нас более 70 лет / розовые , бордовые и белые / на фото цветы 2018г / п .

In [11]:
assert X_train[10][1] == 'продам иж планета 3 , 76 год , ( стоит на старом учёте , документы утеряны ) на ходу , хорошее состояние , все интересующие вопросы по телефону ( с родной коляской на 3 тысячи дороже ) . торга не будет .'

### BOW (1.5 балла)

Один из традиционных подходов -- построение bag of words.

Метод состоит в следующем:

 - Составить словарь самых часто встречающихся слов
 - Для каждого примера посчитать, сколько раз каждое слово из словаря в нём встречается


В sklearn есть CountVectorizer, но в этом задании его использовать нельзя.

__Задание:__ Найдите 10000 самых частых слов из `title` и `description` обучающей выборки, отсортируйте их по убыванию частотности.

In [13]:
# Your code here

all_words = np.concatenate([[j for j in i.split()] for i in tqdm(np.concatenate(X_train))])

Voc, Cnt = np.unique(all_words, return_counts=True)
bow_vocabulary = Voc[np.argsort(Cnt)][::-1][:10000]
bow_vocabulary_cnt = Cnt[np.argsort(Cnt)][::-1][:10000]

In [14]:
# Проверка
sorted(bow_vocabulary)[::200] 

['!',
 '128',
 '2800',
 '71',
 'c5',
 'gigabyte',
 'michelin',
 'solaris',
 '»:/',
 'атлант',
 'болеро',
 'верхней',
 'воспользовавшись',
 'г',
 'грыж',
 'джили',
 'доступны',
 'задач',
 'зимних',
 'использованию',
 'кд',
 'компания',
 'красивые',
 'левое',
 'макс',
 'мех',
 'музей',
 'натяжной',
 'носили',
 'однокомнатная',
 'ответственный',
 'пассажиров',
 'плотная',
 'покрытие',
 'посуточно',
 'приобретения',
 'профессиональные',
 'разрешение',
 'решетку',
 'сб',
 'синее',
 'создание',
 'стандартные',
 'счету',
 'тонировка',
 'удобна',
 'фару',
 'хюндай',
 'шерстяной',
 'этого']

Я не понимаю, как нужно отсортировать bow_vocabulary, чтобы в точности assert получить...

In [15]:
assert sorted(bow_vocabulary)[::200] == ['!', '12500', '270', '700', 'by', 'gh', 'michael', 'sonata', 'ø', 'аудиоподготовка', 'большим', 'веса', 'воспроизведения', 'габариты', 'гтд', 'джинсами', 'доступность', 'загрузки', 'зимней', 'использовался', 'квартала', 'коммуникации', 'кошки', 'лакированные', 'магазин', 'металл', 'мск', 'натуральным', 'носке', 'одному', 'отвечаем', 'пассат', 'плотно', 'покраску', 'постоянные', 'примеры', 'просьба', 'размещайте', 'репетитор', 'сантехник', 'сидения', 'современного', 'стала', 'схема', 'тон', 'удлиненная', 'фасад', 'цветами', 'шея', 'эту']

AssertionError: 

In [16]:
def text_to_bow(text: str) -> np.array:
    """
    Возвращает вектор, где для каждого слова из bow_vocabulary
    указано количество его употреблений
    """ 
    # Your code here
    return np.sum([(bow_vocabulary == i).astype('int') for i in text.split()], axis=0)
        

# Проверка
np.where(text_to_bow("сдаётся уютный , тёплый гараж для стартапов в ml") != 0)[0]

array([   1,    4,   12,  565,  866, 1598, 2533, 4117])

In [17]:
assert np.allclose(np.where(text_to_bow("сдаётся уютный , тёплый гараж для стартапов в ml") != 0)[0],
                   np.array([   1,    4,   12,  565,  866, 1601, 2539, 4063])
)

AssertionError: 

In [18]:
def items_to_bow(items: np.array, use_title=False) -> np.array:
    """ Для каждого товара возвращает вектор его bow """
    # Давайте строить bow только из description товара
    
    # Your code here
    if use_title:
        return np.array([text_to_bow(i) for i in tqdm(items)])
    else:
        return np.array([text_to_bow(i[1]) for i in tqdm(items)])

In [19]:
# Проверка
np.where(items_to_bow([X_train[42]])[0] != 0)

(array([   0,    1,    2,    5,    6,    7,   12,   27,   41,   49,  110,
         189,  207,  221, 2037, 3075, 7307, 9819]),)

In [20]:
assert np.allclose(np.where(items_to_bow([X_train[42]])[0] != 0),
                   np.array([   0, 1, 2, 5, 6, 7, 12, 27, 41, 49, 110,
                                189,  208,  221, 2032, 3052, 7179, 9568]),
)

AssertionError: 

In [21]:
X_train_bow = items_to_bow(X_train)
X_test_bow = items_to_bow(X_test)

### Логистическая регрессия и SVC (1 балл)


Теперь описание каждого товара представлено, как точка в многомерном пространстве.
Очень важно запомнить эту идею: дальше мы будем рассматривать разные способы перехода от текста к точке в пространстве.

Для BOW каждое измерение в пространстве -- какое-то слово.
Мы предполагаем, что текст описывается набором каких-то популярных слов, которые в нём встречаются, а близкие по смыслу тексты будут использовать одинаковые слова.

Обучите логистическую регрессию и SVC с базовыми параметрами.

**Подсказка: рассмотрите использование sparse-матриц - в них хранятся только ненулевые элементы.** `scipy.sparse.csr_matrix(X_train_bow)`

In [12]:
from scipy import sparse
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC

In [22]:
# Sparse матрицы
X_train_sparsed = sparse.csr_matrix(X_train_bow)
X_test_sparsed = sparse.csr_matrix(X_test_bow)

In [23]:
# Логистическая регрессия
bow_model_lr = LogisticRegression(max_iter=100)
bow_model_lr.fit(X_train_sparsed, y_train)
acc_lr_bow = accuracy_score(bow_model_lr.predict(X_test_sparsed), y_test)
print(acc_lr_bow)

assert acc_lr_bow > 0.7

0.5422222222222223


/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


AssertionError: 

In [24]:
# SVC
bow_model_svc = LinearSVC(max_iter=70)
bow_model_svc.fit(X_train_sparsed, y_train)
acc_svc_bow = accuracy_score(bow_model_svc.predict(X_test_sparsed), y_test)
print(acc_svc_bow)

assert acc_svc_bow > 0.68

0.5166666666666667


/opt/conda/lib/python3.6/site-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


AssertionError: 

Accuracy чуть выше 50% говорит об неудовлетворительном качестве предсказаний моделей. 

### Модификация признаков (0.5 балла)

Постройте признаковое пространство на конкатенации (через пробел) `title` и `description`.

In [14]:
# Your code here

# Конкатенация
X_train_concat = np.array([' '.join(i) for i in X_train])
X_test_concat = np.array([' '.join(i) for i in X_test])

In [26]:
# BoW представление
X_train_bow = items_to_bow(X_train_concat, use_title=True)
X_test_bow = items_to_bow(X_test_concat, use_title=True)

# sparse матрицы
X_train_sparsed = sparse.csr_matrix(X_train_bow)
X_test_sparsed = sparse.csr_matrix(X_test_bow)

In [28]:
# Проверка
X_train_bow[0]

array([0, 0, 0, ..., 0, 0, 0])

In [29]:
# Логистическая регрессия
bow_model_lr = LogisticRegression(max_iter=100)
bow_model_lr.fit(X_train_sparsed, y_train)
acc_lr_bow = accuracy_score(bow_model_lr.predict(X_test_sparsed), y_test)
print('accuracyLR =', acc_lr_bow)

accuracyLR = 0.5851111111111111


/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [30]:
# SVC
bow_model_svc = LinearSVC(max_iter=70)
bow_model_svc.fit(X_train_sparsed, y_train)
acc_svc_bow = accuracy_score(bow_model_svc.predict(X_test_sparsed), y_test)
print('accuracySVC =', acc_svc_bow)

accuracySVC = 0.5344444444444445


/opt/conda/lib/python3.6/site-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Добавление title слегка увеличило качество моделей. Но оно все еще неудовлетворительно.

### mystem (1.5) балла

Как можно заметить, в текстах одни и те же слова могут быть в разных падежах, а соответственно в bow это будут разные признаки. 
Чтобы исправить это, можно лемматизировать слова - с помощью библиотеки Mystem.

In [31]:
!pip install pymystem3

In [32]:
from pymystem3 import Mystem

m = Mystem()
m.lemmatize("Здравствуйте, ваше благородие")

Installing mystem to /root/.local/bin/mystem from http://download.cdn.yandex.net/mystem/mystem-3.1-linux-64bit.tar.gz


['здравствовать', ', ', 'ваш', ' ', 'благородие', '\n']

__Задание:__ 
* Лемматезируйте `title` и `description` с помощью mystem
* Gостройте BoW-представление
* Обучите на нём логистическую регрессию и SVC.

In [33]:
# Your code here

# Лемматизация
all_words = np.concatenate([m.lemmatize(X_train_concat[i]) for i in tqdm(range(len(X_train_concat)))])
all_words[:10]

array(['сапог', ' ', '46', ' ', 'размер', ' ', 'новый', ' ', 'сапог', ' '],
      dtype='<U250')

In [35]:
# BoW-словарь
Voc, Cnt = np.unique(all_words, return_counts=True)
bow_vocabulary = Voc[np.argsort(Cnt)][::-1][:10000]
bow_vocabulary_cnt = Cnt[np.argsort(Cnt)][::-1][:10000]

In [36]:
def text_to_lem(text: str) -> np.array:
    """
    Возвращает вектор, где для каждого слова из bow_vocabulary
    указано количество его употреблений
    """
    
    return np.sum([(bow_vocabulary == i).astype('int') for i in m.lemmatize(text)], axis=0)

In [38]:
def items_to_lem(items: np.array) -> np.array:
    """ Для каждого товара возвращает вектор его bow """
    
    return np.array([text_to_lem(i) for i in tqdm(items)])

In [39]:
# BoW представление
X_train_bow = items_to_lem(X_train_concat)
X_test_bow = items_to_lem(X_test_concat)

# Sparse матрицы
X_train_sparsed = sparse.csr_matrix(X_train_bow)
X_test_sparsed = sparse.csr_matrix(X_test_bow)

In [40]:
# Логистическая регрессия
bow_model_lr = LogisticRegression(max_iter=100)
bow_model_lr.fit(X_train_sparsed, y_train)
acc_lr_bow = accuracy_score(bow_model_lr.predict(X_test_sparsed), y_test)
print('accuracyLR =', acc_lr_bow)

accuracyLR = 0.5266666666666666


/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [41]:
# SVC
bow_model_svc = LinearSVC(max_iter=70)
bow_model_svc.fit(X_train_sparsed, y_train)
acc_svc_bow = accuracy_score(bow_model_svc.predict(X_test_sparsed), y_test)
print('accuracySVC =', acc_svc_bow)

accuracySVC = 0.664


/opt/conda/lib/python3.6/site-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Можно видеть, что после лемматизации качество SVC значительно улучшилось. В то же время, качество линейной регрессии незначительно уменьшилось.

66% для SVC -- уже удовлетворительный результат.

### TF-IDF (1.5 балла)

Не все слова полезны одинаково, давайте попробуем [взвесить](http://tfidf.com/) их, чтобы отобрать более полезные.


> TF(t) = (Number of times term t appears in a document) / (Total number of terms in the document).
> 
> IDF(t) = log_e(Total number of documents / Number of documents with term t in it).


В sklearn есть TfidfVectorizer, но в этом задании его использовать нельзя.

Давайте для простоты считать один tf-idf для title и description. Для каждого слова из bow_vocabulary нужно посчитать:
1. Сколько раз оно встретилось в title и description во всём X_train
2. В тексте скольких товаров встретилось это слово

In [42]:
ndoc = len(X_train_concat)
# tf-idf для каждого слова из bow_vocabulary
idf_cnt = np.nan_to_num(
    np.array([np.log(ndoc/np.sum([1 for j in X_train_concat if i in j])) for i in tqdm(bow_vocabulary)]),
    nan=0, posinf=0, neginf=0)

# Проверка
print(bow_vocabulary[:8])
print(idf_cnt[:8])

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: RuntimeWarning: divide by zero encountered in double_scalars
  This is separate from the ipykernel package so we can avoid doing imports until



[' ' '.' ' , ' ' / ' ' - ' 'в' 'и' '\n']
[0.         0.25053922 0.36192681 0.81246711 0.78484485 0.08198507
 0.04111966 0.        ]


In [44]:
def text_to_tfidf(text: str) -> np.array:
    """
    Возвращает вектор, где для каждого слова из bow_vocabulary указан tf-idf
    """

    # Your code here
    text_bow = np.sum([(bow_vocabulary == i).astype('int') for i in m.lemmatize(text)], axis=0)
    return np.nan_to_num(text_bow / np.sum(text_bow), nan=0, posinf=0, neginf=0) * idf_cnt


print(bow_vocabulary[np.where(text_to_tfidf(X_train_concat[0]) != 0)[0]])

['новый' 'размер' '46' 'сапог']


In [46]:
def items_to_tfidf(items: np.array) -> np.array:
    return np.array([text_to_tfidf(i) for i in tqdm(items)])

### Модели на TF-IDF признаках (1 балл)

Обучите логистическую регрессию и SVC, оцените качество (accuracy_score)

In [47]:
# Your code here

# tf-idf представление
X_train_bow = items_to_tfidf(X_train_concat)
X_test_bow = items_to_tfidf(X_test_concat)

# Sparse матрицы
X_train_sparsed = sparse.csr_matrix(X_train_bow)
X_test_sparsed = sparse.csr_matrix(X_test_bow)

In [48]:
# Логистическая регрессия
bow_model_lr = LogisticRegression(max_iter=100)
bow_model_lr.fit(X_train_sparsed, y_train)
acc_lr_bow = accuracy_score(bow_model_lr.predict(X_test_sparsed), y_test)
print('accuracyLR =', acc_lr_bow)

accuracyLR = 0.642


/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [49]:
# SVC
bow_model_svc = LinearSVC(max_iter=70)
bow_model_svc.fit(X_train_sparsed, y_train)
acc_svc_bow = accuracy_score(bow_model_svc.predict(X_test_sparsed), y_test)
print('accuracySVC =', acc_svc_bow)

accuracySVC = 0.7941111111111111


/opt/conda/lib/python3.6/site-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Благодаря tf-idf взвешиванию удалось значительно увеличить качество моделей. 

Для логистической регрессии доля правильных ответов достигла 64%

Для SVC -- 79%

### Hashing Vectorizer (0.5 балла)

Попробуйте использовать `sklearn.feature_extraction.text.HashingVectorizer` для векторизации текстов.
Обязательно оцените качество работы алгоритмов классификации с использованием новой векторизации.

In [24]:
# Your code here
from sklearn.feature_extraction.text import HashingVectorizer
hash_vec = HashingVectorizer()
X_train_sparsed = hash_vec.fit_transform(X_train_concat)
X_test_sparsed = hash_vec.transform(X_test_concat)

In [25]:
print(X_train_sparsed.shape, X_test_sparsed.shape)

(21000, 1048576) (9000, 1048576)


In [26]:
# Логистическая регрессия
bow_model_lr = LogisticRegression(max_iter=100)
bow_model_lr.fit(X_train_sparsed, y_train)
acc_lr_bow = accuracy_score(bow_model_lr.predict(X_test_sparsed), y_test)
print('accuracyLR =', acc_lr_bow)

c:\users\nikir\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
c:\users\nikir\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


accuracyLR = 0.7236666666666667


In [27]:
# SVC
bow_model_svc = LinearSVC(max_iter=70)
bow_model_svc.fit(X_train_sparsed, y_train)
acc_svc_bow = accuracy_score(bow_model_svc.predict(X_test_sparsed), y_test)
print('accuracySVC =', acc_svc_bow)

accuracySVC = 0.8245555555555556


Векторизация с hashing еще сильнее улучшила качество моделей (да еще и очень быстро). Судя по всему, сказывается использование большого числа признаков -- 1048576

Качество логистической регрессии достигло 72%, тогда как метод опорных векторов дает результат в 82% правильных ответов

### Word Vectors (1.5 балл)

Давайте попробуем другой подход -- кажому слову сопоставим какой-то эмбеддинг (вектор).

Вектора будут небольшой размерности. Таким образом мы снизим количество параметров в модели.

Вектора мы возьмём уже готовые (обученные на текстах их интернета), так что наша модель будет знать некоторую дополнительную информацию о внешнем мире.

## См. файл Razuvaev_N_HW3_p2.ipynb 

### Что дальше?

Решение каждого пункта 2 балла:



1. N-Gram модели текстовой классификации
    * Признаки - mystem
    * n-gramm'ы - несколько слов подряд объединяются в один токен
    * Модели - только логистическая регрессия
    * Настоятельно рекомендуется использовать sparse-матрицы
    

2. Использовать Vowpal Wabbit вместо sklearn
    * Признаки - обычный BoW title+description 
    * Главный вызов - заставить работать библиотеку и подготовить признаки

3. Другие способы лемматизации (pymorphy2, spaCy)


* Снабжайте код пояснениями;
* Обязательно необходимо написать вывод по каждому пункту, который вы реализуете.